src/identity-utils.ts

```diff
@@ -0,0 +1,160 @@
+import { LocalStorageVault } from './vault';
+import Types from './types';
+import { IdentityApiData, UserIdentities } from '@mparticle/web-sdk';
+import { isObject } from './utils';
+import { MParticleWebSDK } from './sdkRuntimeModels';
+
+export interface IKnownIdentities extends UserIdentities {
+    device_application_stamp?: string;
+}
+
+export interface ICachedIdentityCall extends UserIdentities {
+    responseText: string;
+    status: number;
+    expireTimestamp: number;
+}
+
+export const cacheIdentityRequest = (
+    method: string,
+    identities: IKnownIdentities,
+    expireTimestamp: number,
+    idCache: LocalStorageVault<Dictionary>,
+    xhr: XMLHttpRequest
+): void => {
+    let cache: Dictionary<ICachedIdentityCall> = idCache.retrieve() || {};
+    let cacheKey = concatenateIdentities(method, identities);
+
+    cache[cacheKey] = { responseText: xhr.responseText, status: xhr.status, expireTimestamp};
+    idCache.store(cache);
+};
+
+// We need to ensure that identities are concatenated in a deterministic way, so
+// we sort the identities based on their enum.
+// we create an array, set the user identity at the index of the user identity type
+export const concatenateIdentities = (
+    method: string,
+    userIdentities: IKnownIdentities
+): string => {
+    // set DAS first since it is not an official identity type
+    let cacheKey: string = `${method}:device_application_stamp=${userIdentities.device_application_stamp};`;
+    const idLength: number = Object.keys(userIdentities).length;
+    let concatenatedIdentities: string = '';
+
+    if (idLength) {
+        let userIDArray: Array<string> = new Array(idLength);
+        // create an array where each index is equal to the user identity type
+        for (let key in userIdentities) {
+            if (key === 'device_application_stamp') {
+                continue;
+            } else {
+                userIDArray[Types.IdentityType.getIdentityType(key)] =
+                    userIdentities[key];
+            }
+        }
+
+        concatenatedIdentities = userIDArray.reduce(
+            (prevValue: string, currentValue: string, index: number) => {
+                let idName: string = Types.IdentityType.getIdentityName(index);
+                return `${prevValue}${idName}=${currentValue};`;
+            },
+            cacheKey
+        );
+    }
+
+    return concatenatedIdentities;
+};
+
+export const hasValidCachedIdentity = (
+    method: string,
+    proposedUserIdentities: IKnownIdentities,
+    idCache?: LocalStorageVault<Dictionary>
+): Boolean => {
+    // There is an edhge case where multiple identity calls are taking place 
+    // before identify fires, so there may not be a cache.  See what happens when 
+    // the ? in idCache is removed to the following test
+    // "queued events contain login mpid instead of identify mpid when calling 
+    // login immediately after mParticle initializes"
+    const cache = idCache?.retrieve();
```

Comment:
Instead of taking in the entire Vault Object, maybe we can just have it take the output of the `Vault.retrieve()`, and return a boolean if the item exists?

This also might help remove some of the need for the guards below.

Reply:
Can you clarify?  It's been a month since this message.  Do you still feel strongly about this?

Reply:
My initial feedback was that instead of having the third argument be `BaseVault<Dictionary>` it should be an the output of `idCache.retrieve()`. This makes the cache functionality less dependent on the storage method (vault versus some in memory) versus the actual data. I think it only saves us one line of code, so maybe it's not a big deal to leave this as is. I think this is self-contained enough where we can clean it up later if need be.